src/dataflow/src/render/join/delta_join.rs

```diff
@@ -304,42 +304,59 @@ where
                         use timely::dataflow::operators::Map;
 
                         // Ensure this input is rendered, and extract its update stream.
-                        let mut update_stream = if let Some((_key, val)) = arrangements
+                        let update_stream = if let Some((_key, val)) = arrangements
                             .iter()
                             .find(|(key, _val)| key.0 == &source_relation)
                         {
                             let as_of = self.as_of_frontier.clone();
                             match val {
-                                Ok(local) => build_update_stream(local, as_of, source_relation)
-                                    .enter_region(region),
-                                Err(trace) => build_update_stream(trace, as_of, source_relation)
-                                    .enter_region(region),
+                                Ok(local) => {
+                                    let (update_stream, err_stream) = build_update_stream(
+                                        local,
+                                        as_of,
+                                        source_relation,
+                                        initial_closure,
+                                    );
+                                    inner_errs.push(err_stream);
+                                    update_stream.enter_region(region)
+                                }
+                                Err(trace) => {
+                                    let (update_stream, err_stream) = build_update_stream(
+                                        trace,
+                                        as_of,
+                                        source_relation,
+                                        initial_closure,
+                                    );
+                                    inner_errs.push(err_stream);
+                                    update_stream.enter_region(region)
+                                }
                             }
                         } else {
-                            self.collection(&inputs[source_relation])
+                            let mut update_stream = self
```

Comment:
should we not panic in this case?

Reply:
We generally prefer to not panic if we can help it because if Materialize crashes it is very disruptive to the user. 

If we have reached this case, then the optimizer, specifically the transform `JoinImplementation`, has made a mistake, and our plan may be suboptimal. But it is totally possible to render the dataflow, so we do.

Perhaps we should put here:
* a comment that if we have entered this code, a mistake has been made upstream.
* a warning to be sent to the logs so we (and users) can see if we ever enter this code path.

Reply:
Right, that's reasonable. However, that leaves us with a code path that is not possible to exercise reliably and hence, test properly. Perhaps, we could add an optimizer hint to force the physical plan for the join, so we can force a delta query even in the absence of the required arrangements.